### DISCUSSION TITLE
miss-readings with values 2, 3, 5, 7

### DISCUSSION THREAD

**Discussion by schneich** [jomjol/AI-on-the-edge-device#2423] [Q&A] [answered]
Hi,
my ESP32 is now running for 3 days. Overall, this is running very smoothly. Thank you all for creating this project!

Unfortunately, I get some miss-readings. The AI often confuses the numbers 2, 3, 5 and 7, when in transition in-between. Think of the little wheel turning from "2" to "3": the lower "dash" of the "2" combined with the upper part of "3" could be confused with a "5".

My current _Model_ setting is `dig-class100_0160_s2_q.tflite`.

The miss-readings in the lower numbers is only a couple of litres -> not an issue, but if this happens in higher numbers as well, it will read 500 litres instead of 300 litres, which is a big difference for our daily water consumption.

I found the description of the models in the [documentation](https://jomjol.github.io/AI-on-the-edge-device-docs/Choosing-the-Model/).
![grafik](https://github.com/jomjol/AI-on-the-edge-device/assets/84641760/76f6c95b-6961-4ca4-8746-a54622c8ac64)

Is there a best practice on how to reduce miss-readings? Which Model are you using? (To be on the safe side, I'd rather have too low readings, then too high readings.)

thank you very much,
Chris

p.s.: please find attached today's log: [AIotE_log.csv](https://github.com/jomjol/AI-on-the-edge-device/files/11524241/AIotE_log.csv)

p.s.: I just had another look at the documentation and changed the model to `dig-cont_0611_s3_q.tflite`. Should this improve the readings?

**Comment by caco3** [comment]
Try the different models and use he one that fits you best.

Please provide a screenshot of your meter.
Did you make sure your digits are contained in the training data? Check the link shown below.
Fals reading can be detected with the neg rate and max rate parameters.

**Reply by schneich** [reply]
Hi @caco3,

thank you. I believe, my number-type is already there.

![grafik](https://github.com/jomjol/AI-on-the-edge-device/assets/84641760/a6363bcc-0947-4cdd-895c-1deeef356a4e)

Chris

**Comment by github-actions** [comment]
See [Digital Digits](https://jomjol.github.io/neural-network-digital-counter-readout) resp. [Analogue Pointers](https://jomjol.github.io/neural-network-analog-needle-readout) for an overview of all trained data.
If your type is not contained it can be added to our training material, see [here](https://jomjol.github.io/AI-on-the-edge-device-docs/collect-new-images/).

**Comment by caco3** [comment]
You have some reflection on the digits, please try to improve this.

Afterwards i would suggest to provide some training date. Indeed there are some number is in the training data, but more of your type will improve the quality.

**Reply by ghost** [reply]
The reflections are not much. Should not be the reason for false predictions.

**Comment by schneich** [comment]
In order to get rid of the refection, I need to re-design the case and 3D-print it again. This will take some time...
As soon as done, I am happy to share some training data.

**Comment by ghost** [comment]
Look at https://jomjol.github.io/AI-on-the-edge-device-docs/ROI-Configuration/

Your ROIs are to small


![](https://user-images.githubusercontent.com/412645/199028748-c48ef5bb-a8d4-4c77-9faf-763e6cf77351.png)

**Reply by schneich** [reply]
Is there a way to change the ratio between inner and outer rectangle? If I draw the rectangle and the inner rectangle is okay, then the outer rectangle is to large - or if the outer rectangle fits the digit window, the inner rectangle is too small.
If this is a questions of pixel, a zoom-functionality would be helpful. :-)

**Reply by ghost** [reply]
Uncheck "Lock Aspect Ratio"

**Reply by ghost** [reply]
For Zooming I use the browser functionality.

**Reply by schneich** [reply]
Sorry for the late response. I had to give some love to another project...

I re-created my casing and I re-configured my AI on the edge project:

![grafik](https://github.com/jomjol/AI-on-the-edge-device/assets/84641760/a6ea3520-6e3a-410c-a696-85c676bec432)


It seems as if the readings are now more or less okay. Over the last 2 days, almost no miss readings. Thanks for the hint to enlarge the ROIs.

Please consider my screenshot: is it worth providing my "Recognized Digit ROIs" to train your model? If so, I am happy to share them.

**Reply by ghost** [reply]
New data are always welcome. But please share your ROIs Configuration (Screenshot) before.

I meant you should make the ROIs wider, not longer.

It's not easy to label ROIs larger than the digit in positions x.8-x.2, because otherwise the scale lines would have to be compressed.

**Reply by schneich** [reply]
Hi @haverland,

Thank you for your reply. I tried to get the inner rectangle closely around the figures. Too close?

>It's not easy to label ROIs larger than the digit in positions x.8-x.2, because otherwise the scale lines would have to be compressed.

I am not sure, if I fully understand this. Could you please elaborate a bit?
In your screenshot above, it says that the outer rectangle should fit the digit window. The inner rectangle should fit around the digit. Well I can do one of this, not both. My first strategy was to get the outer rectangle onto the digit window. This didn't work very well. Now, as I tried to fit the digits into the inner rectangle, it seems to work better. There still are a couple of errors during the day, but nothing of impact to my view in home assistant.


Here my ROIs configurations:
![grafik](https://github.com/jomjol/AI-on-the-edge-device/assets/84641760/f6d6de32-b969-4812-bf03-6fb3fc342310)

![grafik](https://github.com/jomjol/AI-on-the-edge-device/assets/84641760/e524358a-8f0f-4f29-9cc4-f5401e7a7873)



Should I make the ROIs even larger on both, digital and analog ROIs?

**Reply by ghost** [reply]
The digits I would use delta x = 21, delta y = 48

The analog pointer I would change to smaller, because the picture will be reduced to 32x32 pixel, before it used for the neuronal network as input. The red pointer is very thin if it will scaled down. The numbers of the analog pointer have very little influence on the result of the neural network.
A bit more pixels than the black circle should be enough.

**Reply by schneich** [reply]
So these are the new settings. The analog read seems to be problematic, as the camera is in an angle to it.

![grafik](https://github.com/jomjol/AI-on-the-edge-device/assets/84641760/27f1f718-39f8-4ae3-9c3c-8545aa2c1645)

If I start gathering the data: do I need to provide only the "full digits" or the intermediate states as well? (1.0, 1.1, 1.2, 1.3, ..., 1.8, 1.9, 2.0, 2.1, ... -> do you train the transition between two full digits as well?

**Reply by ghost** [reply]
The analog a bit bigger. The lines to the numbers should be in the ROI. 

Digits are ok.

**Reply by ghost** [reply]
You could add more saturation in config, to get more red pointer.

**Comment by schneich** [comment, marked as answer]
Dear all,

just wanted to let you know: in the last couple of months, I changed one or two things in my watermeter. What did the trick in the end: I added a fibre-optic-cable to better distribute the light (https://github.com/jomjol/AI-on-the-edge-device/discussions/1949#discussioncomment-7355535). This reduced the contrast a little, but reduced the reflections. I have not had any misreadings for a couple of days now.

Chris
